README
CPAN-Tester-WWW-Blog
====================

This distribution contains the files used to run and manage the CPAN Testers 
Blog website.

The CPAN Testers Blog website is built on top of the Labyrinth Website 
Management System. As such, the code to drive the website is conatined within
the Labyrinth and associated plugin prerequisites. In order to define the 
website in terms of functionality, layout and style, the files within the
'./vhost' directory of this distribution should be installed into your web
server's virtual host directory.

Included in the distribution is a C<vhost.conf> file, which contains the 
virtual host settings to implement the site using the Apache Web Server.

INSTALLATION

This distribution is in two parts, the first is the installer and the second
are the website files.

In the first instance the installer will install Labyrinth and associated 
plugins. These library modules form the engine of the website. These can be
installed as

  perl Makefile.PL
  make
  make test
  make install

In the second instance, download the latest tarball and copy the contents of
the 'vhost' directory into the directory you plan to use to provide access to
the Blog. Create an appropriate database in MySQL and load the contents of the
SQL file, ./cgi-bin/db/cpanblog-base.sql. Note that this only provides the 
base tables and contents, none of the current CPAN Testers Blog posts. 

You then need to update the configuration file, ./cgi-bin/config/settings.ini,
with the necessary database connection settings, together with the correct
paths to your virtual directory. Then you'll need update the run script,
./cgi-bin/pages.cgi, to point your configuration file.

Finally, you'll need to update your web server as necessary. The CPAN Testers
Blog runs under Apache2, and the vhost file, ./vhosts.conf, can be adapted as
appropriate for other web servers applications.

CPAN TESTERS FUND

CPAN Testers wouldn't exist without the help and support of the Perl 
community. However, since 2008 CPAN Testers has grown far beyond the 
expectations of it's original creators. As a consequence it now requires
considerable funding to help support the infrastructure.

In early 2012 the Enlightened Perl Organisation very kindly set-up a
CPAN Testers Fund within their donatation structure, to help the project
cover the costs of servers and services.

If you would like to donate to the CPAN Testers Fund, please follow the link
below to the Enlightened Perl Organisation's donation site.

https://members.enlightenedperl.org/drupal/donate-cpan-testers

If your company would like to support us, you can donate financially via the
fund link above, or if you have servers or services that we might use, please
send an email to admin@cpantesters.org with details.

Our full list of current sponsors can be found at our I <3 CPAN Testers site.

http://iheart.cpantesters.org

COPYRIGHT AND LICENCE

Copyright (C) 2008-2014 Barbie for Miss Barbell Productions.

This distribution is free software; you can redistribute it and/or modify it 
under the Artistic License v2.

ADDITIONAL COPYRIGHT

This distribution includes TinyMCE [1], a Javascript WYSIWYG Editor used within
the site. TinyMCE is distributed under the GNU Lesser General Public License 
Version 2.1, and all files within the './vhost/html/js/tiny_mce' directory
are copyright of the TinyMCE project.

[1] http://tinymce.moxiecode.com/


Changes
Changes for the CPAN Testers Blog website
=========================================

        - added CSS green ribbon promotion.

1.12    2015-10-11
        - updates from live
        - upgrade to Labyrinth-Plugin-CPAN-3.55.
        - added CPAN Testers Sponsor adverts.
        - added 1 <3 CPAN Testers framed ribbon.

1.11    2015-09-15
        - upgrade to Labyrinth-5.32.
        - upgrade to Labyrinth-Plugin-Core-5.19.

1.10    2015-02-09
        - upgrade to Labyrinth-5.24.
        - removed Perl promotions links.
        - added Admin site links.

1.09    2014-09-21
        - fixed license fields in META.json to be lists.
        - added documentation about the CPAN Testers Fund.

1.08    2014-05-23
        - upgrade to Labyrinth-5.21.
        - upgrade to Labyrinth-Plugin-Core-5.15.
        - upgrade to Labyrinth-Plugin-Articles-Diary 1.05.
        - upgrade to Labyrinth-Plugin-Event 1.04.
        - fixed distribution name in META.
        - added LICENSE file.
        - added nofollow link attributes to iheart.js.
        - upgraded comment management.
        - extend allowable CMS HTML tags.

1.07    2014-02-16
        - removed unused files.
        - template cleanups.
        - upgrade to Labyrinth-5.20.
        - upgrade to Labyrinth-Plugin-Core-5.14.
        - updated sumhits script.

1.06    2013-03-24
        - optimised static images.
        - minimised CSS.
        - added robots.txt.
        - change to JSON for iheart 
        - split posted by and on onto 2 different lines
        - upgrade to Labyrinth 5.15.
        - added block.grango.org to Block/Allow IPs

1.05    2013-01-06
        - change file dates changed to meet W3CDTF standards.
        - added iheart styles.

1.04    2012-04-22
        - fixed META.json test.

1.03    2012-04-22
        - fixed publish status when saving edited comments.
        - upgrade to Labyrinth 5.12.
        - removed social media bar due to dubious practices.
        - added CPAN Testers Sponsors.
        - added README.
        - added minimum perl version (5.006).
        - fixed test pre-requisites.
        - reworked Makefile.PL for clarity.

1.02    2011-08-18
        - upgrade to Labyrinth 5.09.
        - upgrade to Labyrinth-Plugin-Core 5.09.
        - added titlemaps and pathmaps to settings.ini.
        - WAI WCAG v2.0 compliance updates.
        - update layout to include an update notification.
        - added a new social media bar
        - fixed images and css incorrectly pasted from wiki distro
        - revised banner logo image by Jon 'JJ' Allen.

1.01    2011-06-20
        - updates to improve presentation.
        - included a Terms & Conidtions page to meet EU Cookie Directive.
        - updates to diary and user templates for images.
        - updates to templates for usability compliance.
        - fixes for Posted lookups.
        - upgrade to Labyrinth 5.08.
        - upgrade to Labyrinth-Plugin-Core 5.08.

1.00    2011-04-20
        - upgrade to Labyrinth 5.04.
        - upgrade to Labyrinth-Plugin-Core 5.06.
        - upgrade to Labyrinth-Plugin-Articles-Diary 1.00.
        - upgrade to Labyrinth-Plugin-Event 1.00.
        - packaged for CPAN release.

0.06    2010-11-20
        - user lists now hide special users (Guest), restrict access to users
          of the same or higher access, and prevent priveledged users from
          being deleted.
        - user acls fixed to correctly list folders and access.
        - folder list driven by database, added FolderName function.
        - update to admin templates to make them consistent across sites.

0.05    2010-09-29
        - repositioned site boxes to avoid text and logo clashing.

0.04    2010-09-16
        - added links to Metabase and Labyrinth.
        - upgraded to Labyrinth v5.00.

0.03    2010-01-14
        - added quote.html
        - added Analysis to site family lists.

0.02    2009-07-04
        - added TinyMCE support to comments for public users.
        - fixed bug in saving comments for admin users.

0.01    2009-05-17
        - First released version, based on Labyrinth v4.16


vhost/html/js/admin.js
function doSubmit(act,id) {
	document.forms.EDIT.recordid.value = id;
	document.forms.EDIT.doaction.value = act;
	document.forms.EDIT.submit();
}

function doOnlyOne(act) {
	var checkBoxArr = getSelectedCheckbox(document.forms[0].LISTED);
	if (checkBoxArr.length == 1) { 
		doSubmit(act);
		return;
	}

	if (checkBoxArr.length == 0) { alert("No check boxes selected"); }
	else						 { alert("Only one check box can be selected"); }
}


function getSelectedCheckbox(buttonGroup) {
   // Go through all the check boxes. return an array of all the ones
   // that are selected (their position numbers). if no boxes were checked,
   // returned array will be empty (length will be zero)
   var retArr = new Array();
   var lastElement = 0;
   if (buttonGroup[0]) { // if the button group is an array (one check box is not an array)
      for (var i=0; i<buttonGroup.length; i++) {
         if (buttonGroup[i].checked) {
            retArr.length = lastElement;
            retArr[lastElement] = i;
            lastElement++;
         }
      }
   } else { // There is only one check box (it's not an array)
      if (buttonGroup.checked) { // if the one check box is checked
         retArr.length = lastElement;
         retArr[lastElement] = 0; // return zero as the only array value
      }
   }
   return retArr;
} // Ends the "getSelectedCheckbox" function


function SetImage(id,href) {
	f = document.forms[1];
	f.elements[f.display.value].value=id;
	disp = document.getElementById(f.display.value);
	disp.src=webpath+href
}

function PhotoGallery(callback) {
	document.forms[1].display.value = callback;
	window.open(cgipath+"pages.cgi?act=album-pick", "popup", "width=370,height=450,scrollbars=no,menubar=no,resizable=no,status=no,toolbar=no,location=no");
}

function ImageGallery(callback) {
	document.forms[1].display.value = callback;
	window.open(cgipath+"pages.cgi?act=imgs-gallery&amp;imagetype=0,1,2,4,9", "popup", "width=340,height=390,scrollbars=no,menubar=no,resizable=no,status=no,toolbar=no,location=no");
}



vhost/toolkit/rssfeeds.pl
#!/usr/bin/perl -w
use strict;

my $BASE;

# create images in pages
BEGIN {
    $BASE = '/var/www/cpanblog';
}

#----------------------------------------------------------
# Additional Modules

use lib qw|../cgi-bin/lib ../cgi-bin/plugins|;
use Labyrinth::Globals;
use Labyrinth::RSS;
use Labyrinth::Variables;
use Labyrinth::Plugin::Articles;

use File::Basename;
use File::Path;
use IO::File;

#----------------------------------------------------------
# Variables

my @types = (
#    { type => 'rss',  version => '0.9' },
#    { type => 'rss',  version => '1.0' },
    { type => 'rss',  version => '2.0' },
    { type => 'atom', version => '1.0' },
);

#----------------------------------------------------------
# Code

Labyrinth::Globals::LoadSettings("$BASE/cgi-bin/config/settings.ini");
Labyrinth::Globals::DBConnect();

# Diary Entries
$settings{perma} = $tvars{webpath} . '/diary/';
$cgiparams{sectionid} = 6;
$settings{data}{article_limit} = 10;
$settings{data}{article_stop}  = 10;

my $arts = Labyrinth::Plugin::Articles->new();
$arts->List();

$tvars{block} = 'articles/arts-block2.html';

#use Labyrinth::Audit;
#use Data::Dumper;
#LogDebug( Dumper(\%tvars) );

for my $item (@types) {
    my $rss = Labyrinth::RSS->new( %$item, perma => '/diary/' );
    my $xml = $rss->feed(@{$tvars{mainarts}});
    write_xml("rss/$item->{type}-$item->{version}.xml",$xml);
}

sub write_xml {
    my $file = shift;
    my $xml  = shift;
    my $fh = IO::File->new("$BASE/html/$file",'w')    or die "Cannot write to file [$BASE/html/$file]: $!";
    print $fh $xml;
    $fh->close;
}

__END__

=head1 AUTHOR

Barbie, <barbie@missbarbell.co.uk> for
Miss Barbell Productions, L<http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2002-2011 Barbie for Miss Barbell Productions
  All Rights Reserved.

  This module is free software; you can redistribute it and/or
  modify it under the Artistic License 2.0.

=cut


lib/CPAN/Testers/WWW/Blog.pm
package CPAN::Testers::WWW::Blog;

use warnings;
use strict;

our $VERSION = '1.12';

1;

__END__

=head1 NAME

CPAN::Testers::WWW::Blog - The CPAN Testers Blog website.

=head1 DESCRIPTION

This distribution contains all the code used to run the CPAN Testers Blog 
website.

=head1 VHOST FILES

The CPAN Testers Blog website is built on top of the Labyrinth Website 
Management System. As such, the code to drive the website is conatined within
the Labyrinth and associated plugin prerequisites. In order to define the 
website in terms of functionality, layout and style, the files within the
'./vhost' directory of this distribution should be installed into your web
server's virtual host directory.

Included in the distribution is a C<vhost.conf> file, which contains the 
virtual host settings to implement the site using the Apache Web Server.

=head1 ADDITIONAL COPYRIGHT

This distribution includes TinyMCE [1], a Javascript WYSIWYG Editor used within
the site. TinyMCE is distributed under the GNU Lesser General Public License 
Version 2.1, and all files within the './vhost/html/js/tiny_mce' directory
are copyright of the TinyMCE project.

[1] http://tinymce.moxiecode.com/

=head1 SEE ALSO

L<Labyrinth>, 

L<http://blog.cpantesters.org>

=head1 CPAN TESTERS FUND

CPAN Testers wouldn't exist without the help and support of the Perl 
community. However, since 2008 CPAN Testers has grown far beyond the 
expectations of it's original creators. As a consequence it now requires
considerable funding to help support the infrastructure.

In early 2012 the Enlightened Perl Organisation very kindly set-up a
CPAN Testers Fund within their donatation structure, to help the project
cover the costs of servers and services.

If you would like to donate to the CPAN Testers Fund, please follow the link
below to the Enlightened Perl Organisation's donation site.

F<https://members.enlightenedperl.org/drupal/donate-cpan-testers>

If your company would like to support us, you can donate financially via the
fund link above, or if you have servers or services that we might use, please
send an email to admin@cpantesters.org with details.

Our full list of current sponsors can be found at our I <3 CPAN Testers site.

F<http://iheart.cpantesters.org>

=head1 AUTHOR

Barbie, <barbie@missbarbell.co.uk> for
Miss Barbell Productions, L<http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2008-2015 Barbie for Miss Barbell Productions
  All Rights Reserved.

  This module is free software; you can redistribute it and/or
  modify it under the Artistic License 2.0.

=cut


Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'CPAN::Testers::WWW::Blog',
    VERSION_FROM        => 'lib/CPAN/Testers/WWW/Blog.pm',
    ABSTRACT            => 'The CPAN Testers Blog website',
    NO_META             => 1,
    PREREQ_PM => {

        # prereqs
        'Labyrinth'                             => '5.24',
        'Labyrinth::DIUtils::ImageMagick'       => '5.04',
        'Labyrinth::Plugin::Core'               => '5.15',
        'Labyrinth::Plugin::CPAN                => '3.55',
        'Labyrinth::Plugin::Articles::Diary'    => '1.05',
        'Labyrinth::Plugin::Event'              => '1.04',

        # for testing
        'IO::File'                              => 0,       # for testing
        'Test::More'                            => 0.70     # for testing

    },
);
